remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,804 @@
+            // if we access a single non-zero offset, mark just that offset
+            live.set(state_slots[offset].index);
+          } else {
+            live.set();
+          }
+        }
+      }
+    } else if (auto invoke_inst = llvm::dyn_cast<llvm::InvokeInst>(inst)) {
+      for (auto &arg_it : invoke_inst->arg_operands()) {
+        auto arg = arg_it->stripPointerCasts();
+        if (val_to_offset.count(arg)) {
+          auto offset = val_to_offset.at(arg);
+          if (offset != 0) {
+            // if we access a single non-zero offset, mark just that offset
+            live.set(state_slots[offset].index);
+          } else {
+            live.set();
+          }
+        }
+      }
+    } else if (auto store_inst = llvm::dyn_cast<llvm::StoreInst>(inst)) {
+      if (auto scope = GetScopeFromInst(*inst)) {
+        auto inst_size = dl->getTypeAllocSize(store_inst->getOperand(0)->getType());
+        auto &state_slot = state_slots[scope_to_offset.at(scope)];
+        if (!live.test(state_slot.index)) {
+          if (on_remove_pass) {
+            to_remove.push_back(inst);
+          }
+        } else if (inst_size < state_slot.size) {
+          // Partial stores revive the slot.
+          live.set(state_slot.index);
+        } else {
+          // Full stores kill the slot.
+          live.reset(state_slot.index);
+        }
+        // Mark that we have accessed this slot
+        func_used.set(state_slot.index);
+      }
+    } else if (llvm::isa<llvm::LoadInst>(inst)) {
+      if (auto scope = GetScopeFromInst(*inst)) {
+        auto slot_num = state_slots[scope_to_offset.at(scope)].index;
+        // Loads revive the slot.
+        live.set(slot_num);
+        // Mark that we have accessed this slot
+        func_used.set(slot_num);
+      }
+    }
+  }
+  auto &old_live_on_entry = block_map[B];
+  if (old_live_on_entry != live) {
+    old_live_on_entry = live;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+void LiveSetBlockVisitor::PerformRemovePass(bool create_dot = true) {
+  on_remove_pass = true;
+  for (auto &block_it : func) {
+    auto block = &block_it;
+    VisitBlock(block);
+  }
+  if (create_dot) {
+    CreateDOTDigraph();
+  }
+  on_remove_pass = false;
+}
+
+// Remove all dead stores.
+bool LiveSetBlockVisitor::DeleteDeadInsts(void) {
+  bool changed = false;
+  while (!to_remove.empty()) {
+    auto inst = to_remove.back();
+
+    //if (auto *alloc_inst = llvm::dyn_cast<llvm::AllocaInst>(&inst)) {
+      //for (auto *dbg_info_intrinsic : llvm::FindDbgAddrUses(alloc_inst)) {
+        //dbg_info_intrinsic->eraseFromParent();
+      //}
+    //}
+
+    inst->replaceAllUsesWith(llvm::UndefValue::get(inst->getType()));
+
+    for (auto &operand : inst->operands()) {
+      if (auto op_inst = llvm::dyn_cast<llvm::Instruction>(operand)) {
+        operand = nullptr;
+        if (llvm::isInstructionTriviallyDead(op_inst)) {
+          to_remove.insert(to_remove.begin(), op_inst);
```

Comment:
Make this insert at the end of the vector rather than at the beginning. Otherwise the whole thing needs to be continually shifted.